sdk/helper/keysutil/policy.go

```diff
@@ -882,6 +891,87 @@ func (p *Policy) DeriveKey(context, salt []byte, ver int, numBytes int) ([]byte,
 	}
 }
 
+func (p *Policy) DeriveKeyECDH(baseKeyVer int, peerPublicKeyPem []byte, derivedKeySizeInBytes int) ([]byte, error) {
+	var curve elliptic.Curve = elliptic.P256()
+
+	if !p.Type.KeyAgreementSupported() {
+		return nil, errutil.UserError{Err: fmt.Sprintf("base key type %v does not support key agreement", p.Type)}
+	}
+
+	if p.Keys == nil || p.LatestVersion == 0 {
+		return nil, errutil.InternalError{Err: "unable to access the base key; no key versions found"}
+	}
+
+	switch {
+	case baseKeyVer == 0:
+		baseKeyVer = p.LatestVersion
+	case baseKeyVer < 0:
+		return nil, errutil.UserError{Err: "requested version of the base key is negative"}
+	case baseKeyVer > p.LatestVersion:
+		return nil, errutil.UserError{Err: "requested version of the base key is higher than the latest key version"}
+	case baseKeyVer < p.MinAvailableVersion:
+		return nil, errutil.UserError{Err: "requested version of the base key is less than the minimum available key version"}
+	}
+	baseKeyEntry, err := p.safeGetKeyEntry(baseKeyVer)
+	if err != nil {
+		return nil, err
+	}
+
+	switch p.Type {
+	case KeyType_ECDSA_P256:
+		curve = elliptic.P256()
+	case KeyType_ECDSA_P384:
+		curve = elliptic.P384()
+	case KeyType_ECDSA_P521:
+		curve = elliptic.P521()
+	default:
+		return nil, errutil.InternalError{Err: "unsupported base key type for ECDH key agreement"}
+	}
+
+	ownPrivateKey := &ecdsa.PrivateKey{
+		PublicKey: ecdsa.PublicKey{
+			Curve: curve,
+			X:     baseKeyEntry.EC_X,
+			Y:     baseKeyEntry.EC_Y,
+		},
+		D: baseKeyEntry.EC_D,
+	}
+
+	peerPublicKey, err := certutil.ParsePublicKeyPEM(peerPublicKeyPem)
+	if err != nil {
+		return nil, errutil.UserError{Err: fmt.Sprintf("failed to parse peer public key PEM: %s ", err.Error())}
+	}
+
+	if ownPrivateKey.Curve != peerPublicKey.(*ecdsa.PublicKey).Curve {
```

Comment:
With thanks to @fatima2003, this will crash if `peerPublicKey` cannot be safely converted. E.g., if you pass in a RSA public key in the PEM. :-)